Atomic/syscontainers.py

```diff
@@ -2779,6 +2785,89 @@ def _correct_dir_permissions_for_user(path):
                     s = os.stat(fullpath)
                     os.chmod(fullpath, s.st_mode | 0o600)
 
+    def _configure_userns(self, config):
+        """
+        Configure the userns for the specified OCI configuration.
+        """
+        config['linux']['namespaces'] = [i for i in config['linux']['namespaces'] if i['type'] != 'network']
+
+        namespaces = set([i['type'] for i in config['linux']['namespaces']])
+
+        # force the 'user' namespace
+        if 'user' not in namespaces:
+            config['linux']['namespaces'].append({'type' : 'user'})
+            namespaces.add('user')
+
+        all_namespaces = set(['pid', 'uts', 'network', 'mount', 'ipc', 'user'])
+
+        drop_sys = len(all_namespaces - namespaces) > 0
+
+        new_mounts = []
+        for i in config['mounts']:
+            if drop_sys and i['destination'].startswith("/sys"):
+                continue
+            if 'options' not in i:
+                i['options'] = []
+
+            if not i['destination'].startswith('/dev'):
+                if 'nodev' not in i['options']:
+                    i['options'].append('nodev')
+                if 'nosuid' not in i['options']:
+                    i['options'].append('nosuid')
+            new_mounts.append(i)
+
+        if drop_sys:
+            sys_mount = {
+                "destination": "/sys",
+                "type": "bind",
+                "source": "/sys",
+                "options": [
+                    "rbind",
+                    "nosuid",
+                    "noexec",
+                    "nodev",
+                    "ro"
+                ]
+            }
+            new_mounts.append(sys_mount)
+
+        config['mounts'] = new_mounts
+
+        if 'resources' in config['linux'] and 'devices' in config['linux']['resources']:
+            del config['linux']['resources']['devices']
+
+        username = getpass.getuser()
+        def get_subid(root_id, path):
+            prefix = username + ":"
+            r = [
+                {
+                    "containerID": 0,
+                    "hostID": root_id,
+                    "size": 1
+                }
+            ]
+            try:
+                with open(path) as f:
+                    for i in f.readlines():
+                        if i.startswith(prefix):
+                            _, hostid, size = [i.strip() for i in i.split(':')]
+                            extra_ids = {
+                                "containerID": 1,
+                                "hostID": int(hostid),
+                                "size": int(size)
+                            }
+                            r.append(extra_ids)
+
+            except IOError as e:
+                if e.errno != errno.ENOENT:
+                    raise
+            return r
+
+        if 'uidMappings' not in config['linux']:
+            config['linux']['uidMappings'] = get_subid(os.getuid(), "/etc/subuid")
```

Comment:
Question: I guess we add an extra uid/gid mapping here because user might be  also recognized as subordinate uid? I am not too familiar with the concept of subordinate uid here, but do you think it would be nice to add a comment here to explain a bit? :-)

Reply:
the current user is mapped to root in the container

Reply:
right, I think I see that part, but do you mind elaborating a bit about what will happen for the second uid/gid mapping, which is related to the subuid file we input here, that you added conditionally? :-)

Reply:
sure, I try to answer to that :)  runc supports more than an user mapped in the usernamespace, and for that it uses the `newuid/newgid` suid tools for setting it up.  We read the configuration file for these two tools and if a mapping for the current is present, then we add it to the configuration file so that runc can request `newuid/newgid` to set it up.

The same  thing is done by bwrap-oci internally, so we never exposed it into atomic:

https://github.com/projectatomic/bwrap-oci/blob/master/run.c#L730

Reply:
ah gotcha, makes sense.